flake.nix

```diff
@@ -15,51 +15,62 @@
       let
         pkgs = nixpkgs.legacyPackages.${system};
         poetry2nixPkg = poetry2nix.legacyPackages.${system};
+
+        controllerOverrides = final: prev:
+          let
+            addNativeBuildInputs = drvName: inputs: {
+              "${drvName}" = prev.${drvName}.overridePythonAttrs (
+                old: {
+                  nativeBuildInputs =
+                    (old.nativeBuildInputs or [ ]) ++ inputs;
+                }
+              );
+            };
+          in
+          { }
+          // addNativeBuildInputs "mpi4py" [ final.cython ]
+          // addNativeBuildInputs "rexfw" [ final.poetry-core ]
+          // addNativeBuildInputs "chainsail-helpers" [ final.poetry-core ];
+
+        controllerOpts = {
+          projectDir = ./app/controller;
+          overrides = poetry2nixPkg.overrides.withDefaults(controllerOverrides);
+          preferWheels = true;
+        };
+        controller = poetry2nixPkg.mkPoetryApplication controllerOpts;
+        controllerEnv = poetry2nixPkg.mkPoetryEnv controllerOpts;
+
+        basePackages = [
+          pkgs.docker-compose
+          pkgs.file
+          pkgs.gnumake
+          pkgs.kubectl
+          pkgs.kubernetes-helm
+          pkgs.minikube
+          pkgs.ncurses
+          pkgs.nodejs
+          pkgs.openmpi
+          pkgs.poetry
+          pkgs.python38Packages.tkinter
+          pkgs.terraform
+          pkgs.yarn
+        ];
+
+        pythonDevShell = ps: pkgs.mkShell {
+          packages = ps ++ basePackages;
+        };
+        controllerDevShell = pythonDevShell [
+          controllerEnv
+          controller
+        ];
       in
       {
-        devShells.default = pkgs.mkShell {
-          buildInputs = with pkgs; [
-            docker-compose
-            file
-            gnumake
-            kubectl
-            kubernetes-helm
-            minikube
-            ncurses
-            nodejs
-            openmpi
-            poetry
-            python38Packages.tkinter
-            terraform
-            yarn
-            zlib
-          ];
-
-          # Setting the LD_LIBRARY_PATH environment variable.
-          # Can also make use of the `.overrideAttrs` method to prevent from overwriting it.
-          # See PR #310 for details: https://github.com/tweag/chainsail/pull/310.
-          LD_LIBRARY_PATH = "${pkgs.stdenv.cc.cc.lib}/lib:${pkgs.file}/lib:${pkgs.zlib}/lib";
+        devShells = {
+          controller = controllerDevShell;
+          default = pkgs.mkShell { packages = basePackages; };
```

Comment:
It seems a little whimsical to change the definition to `default = pythonDevShell`. The way I see it is that the default environment is different to the ones for working on a specific service/application. i.e. it's not completely symmetrical because there's a distinction. But this is also a bit whimsical on my behalf.

What you suggest doesn't quite work in any case. We'd need to change the argument to `pythonDevShell` to be an attrset and choose a name for the argument in order to give it a default argument. Then the default shell would look like `default = pythonDevShell {}`. However, in the case of getting a working Python environment for doing work on a specific application, an argument is always required (the `mkPoetryEnv` package).

Reply:
I agree with your point that there are quite a few base packages and these could be split up in a later PR. The `pythonDevShell` is a bit of a misnomer, it could be called `completePythonDevShellWithOtherPackagesYouWillNeed` 🙂. The only downside of splitting things up to have less initial downloading is that developers may end up finding that they are in the wrong shell which would be frustrating. We should muse on it a little more and raise an issue if it needs addressing.

Some initial musings:
- A halfway measure may be to pare back some of the dependencies. For instance, I'm not sure what `file` is for.
- If we decide to Nixify the `client` front end, we might find that we can drop `yarn` from the global base packages.

Reply:
Regarding your first message: to be honest, I don't quite follow. https://github.com/tweag/chainsail/blob/controller-dev-shell/flake.nix#L63 tells me that `ps` is a list (because the `++` operator doesn't work for attribute sets. And `pkgs.mkShell { packages = basePackages; };` (which is the current `default`) is exactly the same thing as `pythonDevShell []`. That's why I suggested to replace the current `default` and to rename `pythonDevShell`.

As for splitting up the dependencies, your points are very good. But let's think about those later :slightly_smiling_face:

Reply:
Yes, `ps` is indeed a list. Your original comment would have made more sense to me if you had suggested `default = pythonDevShell []` and at the same time suggest a rename. My perspective was that I wanted to treat the creation of development shells for Python applications as distinct from the one that creates a generic shell with useful utilities. However, I don't think it's important either way.